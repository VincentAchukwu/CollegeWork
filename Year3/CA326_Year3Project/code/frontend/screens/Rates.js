import React, {Component } from "react";
import { StyleSheet, Text, View, TouchableOpacity, StatusBar, FlatList, TextInput, ScrollView } from "react-native";
import LinearGradient from "react-native-linear-gradient";

// specifying URL of Flask server
const flaskAPI = "https://currencyapp200.pythonanywhere.com/api/rates?"

export default class Rates extends Component {
    constructor(props) {
        super(props);
        this.state = {
            base: "",
            date: "",
            rates: "",
            ratesError: null,
            result: null,
            status:"Press",
        };

    };

    // checks if the "currency" field is not empty
    ratesValidator() {
        if(this.state.rates == ""){
            this.setState({ratesError: "Field cannot be empty!"})
        }
        else{
            this.setState({ratesError: ""})
        }
    }

    // Setting up jest button testing
    setStatus() {
        this.setState({
            status: "button pressed"
        })
    }

    render() {
        return (
            <View style={styles.container}>
                <StatusBar backgroundColor="#1E90FF" barStyle="light-content"/>

                <Text style= {styles.footerText}>Base Currency (Default is EUR)</Text>
                <View style= {styles.action}>
                    <TextInput
                        testID="baseCurrency"
                        style={{flex:1}}
                        maxLength={3}
                        placeholder="e.g GBP (optional)"
                        onChangeText={(base) => { this.setState({ base }) }}
                    />
                </View>

                <Text style= {styles.footerText}>Date (Default is latest)</Text>
                <View style= {styles.action}>
                    <TextInput
                        testID="date"
                        style={{flex:1}}
                        maxLength={10}
                        keyboardType={"numeric"}
                        placeholder="YYYY-MM-DD (optional)"
                        onChangeText={(date) => { this.setState({ date }) }}
                    />
                </View>

                <Text style= {styles.footerText}>Enter one or more currencies separated by space</Text>
                <View style= {styles.action}>
                    <TextInput
                        testID="currencyList"
                        style={{flex:1}}
                        onBlur={()=>this.ratesValidator()}
                        placeholder="e.g EUR USD"
                        onChangeText={(rates) => { this.setState({ rates }) }}
                    />
                </View>

                <Text style={styles.error}>{this.state.ratesError}</Text>

                <TouchableOpacity
                    style= {styles.convert}
                    disabled={!this.state.rates}
                    onPress={async () => {
                        const ratesEndoint = "base="+this.state.base+"&date="+this.state.date+"&rates="+this.state.rates;
                        const response = await fetch(flaskAPI+ratesEndoint, {
                            method: "GET",
                            mode: "CORS",
                        })
                        .then(res => res.json())
                        .then(data => {
                            this.setState({ result: data.requestedRates })
                        });
                    }}
                >
                    <LinearGradient
                        colors={ ["#09C6F9", "#045DE9"] }
                        style={ styles.button }
                    >
                        <Text style={styles.convertText}>Get Rates</Text>
                    </LinearGradient>
                </TouchableOpacity>

                <View style = {styles.lineStyle} />

                {/* Button Testing */}
                {/*<TouchableOpacity
                    testID="myButton"
                    onPress= {() => this.setStatus()}
                >
                    <Text>Press</Text>
                </TouchableOpacity>

                <Text testID="myText">{this.state.status}</Text>*/}


                {/* basically the flatlist shows result of each currency rate
                unless the input was invalid, it'll display error message */}
                
                <View style={[styles.action, {marginTop: "1%", height: "20%"}] }>
                    <FlatList
                        data={this.state.result}
                        keyExtractor={item => item.key}
                        renderItem={({ item }) => (
                            <View>
                                {item.key != "invalid" ? <Text>{item.key} = {item.rate}</Text> : <Text style={{color: "red"}}>Please enter valid currency code(s) and date(1999 - Current Date).</Text> }
                            </View>
                        )}
                    />
                </View>
            </View>

        )
    }

}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        alignItems: "center",
        backgroundColor: "#ffffff"
    },
    input: {
        borderWidth: 1,
        padding: 8,
        margin: 10,
        width: 200,
    },
    footerText: {
        fontSize: 18,
        paddingLeft: "3%"
    },
    action: {
        flexDirection: 'row',
        alignItems: 'center',
        borderWidth: .8,
        height: 50,
        borderRadius: 30,
        paddingLeft: "4%",
        paddingRight: "4%",
        margin: "3%",
        width:"50%"
    },
    convert: {
        width: '40%',
        height: 50,
        justifyContent: 'center',
        alignItems: 'center',
        borderRadius: 30,
    },
    button: {
        width: '100%',
        height: 50,
        justifyContent: 'center',
        alignItems: 'center',
        borderRadius: 30
    },
    convertText: {
        fontSize: 18,
        textAlign: "center",
        color: "white",
    },
    error: {
        color: "red",
    },
    lineStyle:{
        borderBottomWidth: 10,
        borderRadius: 30,
        width: "100%",
        paddingTop: "3%",
        paddingBottom: "3%",
        borderBottomColor: "#1E90FF",
    },
})
