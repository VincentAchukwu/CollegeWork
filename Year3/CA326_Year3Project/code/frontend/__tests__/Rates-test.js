import "react-native";
import React from "react";
import Rates from "../screens/Rates";
import renderer from "react-test-renderer";
import {create, act} from "react-test-renderer"

const tree = create(<Rates/>);

// ********* Screen Rendering Test *********

test("Rates Snapshot", () =>{
    const snap = renderer.create(
        <Rates/>
    ).toJSON;

expect(snap).toMatchSnapshot();
});

// *****************************************

// ********* Button Testing *********

test("button press", () => {
    // press button
    const button = tree.root.findByProps({ testID: "myButton"}).props;
    act(() => button.onPress());

    // express text to equal "button pressed"
    const text = tree.root.findByProps({ testID: "myText"}).props;
    expect(text.children).toEqual("button pressed")
});

// ***********************************

// ********* Function to Find Elements in a Tree (Tree being <View/> component and everything wrapped in it) *********

const findElement = function(tree2, element) {
    let result = undefined
    for(node in tree2.children){
        if(tree2.children[node].props.testID = element) {
            result = true;
        }
    }
    return result
};

// *******************************************************************************************************************

// ********* TextInput Tests *********

it("Has Element baseCurrency", () => {
    const tree2 = renderer.create(<Rates/>).toJSON();
    expect(findElement(tree2, "baseCurrency")).toBeDefined();
});

it("Has Element date", () => {
    const tree2 = renderer.create(<Rates/>).toJSON();
    expect(findElement(tree2, "date")).toBeDefined();
});

it("Has Element currencyList", () => {
    const tree2 = renderer.create(<Rates/>).toJSON();
    expect(findElement(tree2, "currencyList")).toBeDefined();
});

// ************************************