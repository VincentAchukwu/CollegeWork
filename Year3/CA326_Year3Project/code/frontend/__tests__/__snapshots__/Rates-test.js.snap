// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rates Snapshot 1`] = `[Function]`;
